Google/google_module/google_drive_reports.py

```diff
@@ -0,0 +1,212 @@
+                    raise
+
+                # compute the duration of the last events fetching
+                duration = int(time.time() - start)
+                FORWARD_EVENTS_DURATION.labels(intake_key=self.configuration.intake_key).observe(duration)
+
+                # Compute the remaining sleeping time
+                delta_sleep = self.configuration.frequency - duration
+                # if greater than 0, sleep
+                if delta_sleep > 0:
+                    time.sleep(delta_sleep)
+
+        finally:
+            self.log(
+                message="Failed to forward events from Google Reports API",
+                level="info",
+            )
+
+    def get_build_object(self):
+        credentials = service_account.Credentials.from_service_account_file(
+            self.service_account_path, scopes=self.scopes
+        )
+        delegated_credentials = credentials.with_subject(self.configuration.admin_mail)
+
+        reports_service = build("admin", "reports_v1", credentials=delegated_credentials)
+
+        return reports_service
+
+    def get_activities(self):
+        reports_service = self.get_build_object()
+
+        self.log(message=f"Start requesting the Google reports with credential object created", level="info")
+
+        activities = (
+            reports_service.activities()
+            .list(
+                userKey="all",
+                applicationName=self.applicationName,
+                maxResults=self.pagination_limit,
+                startTime=self.from_date.strftime("%Y-%m-%dT%H:%M:%SZ"),
+            )
+            .execute()
+        )
+
+        return activities
+
+    def get_next_activities(self, next_key):
+        reports_service = self.get_build_object()
+
+        self.log(message=f"Start requesting the Goole reports with credential object created", level="info")
+
+        activities = (
+            reports_service.activities()
+            .list(
+                userKey="all",
+                applicationName=self.applicationName,
+                maxResults=self.pagination_limit,
+                pageToken=next_key,
+                startTime=self.from_date.strftime("%Y-%m-%dT%H:%M:%SZ"),
+            )
+            .execute()
+        )
+
+        return activities
+
+    def get_next_activities_with_next_key(self, next_key):
+        while next_key:
+            grouped_data = []
+            response_next_page = self.get_next_activities(next_key)
+
+            next_page_items = response_next_page.get("items", [])
+            recent_date = next_page_items[0].get("id").get("time")
+            self.most_recent_date_seen = isoparse(recent_date)
+            grouped_data.extend(next_page_items)
+            self.log(message=f"Sending other batches of {len(grouped_data)} messages", level="info")
+            OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key).inc(len(grouped_data))
+            self.push_events_to_intakes(events=grouped_data)
```

Comment:
```suggestion
            grouped_data.extend(next_page_items)
            self.log(message=f"Sending other batches of {len(grouped_data)} messages", level="info")
            OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key).inc(len(grouped_data))
            self.push_events_to_intakes(events=grouped_data)
            self.most_recent_date_seen = isoparse(recent_date)
```
I would set `most_recent_date_seen` after having sent the events

Reply:
Thanks, Good idea